File: v1/Canteen.go
package v1

import (
	"github.com/julienschmidt/httprouter"
	"munchmate-backend/common"
	"net/http"
	"strconv"
)

type Canteen struct {
	ID          int32
	CityID      int32
	Name        string
	CityName    string
	GeoLocation common.Location
	Distance    float64
}

func GetNearCanteens(w http.ResponseWriter, r *http.Request, par httprouter.Params) {
	// obtain http-get parameters
	v := r.URL.Query()
	lat := v.Get("lat")
	lng := v.Get("lng")

	// try to parse parameters as floating point
	_, latErr := strconv.ParseFloat(lat, 32)
	_, lngErr := strconv.ParseFloat(lng, 32)

	// if parsing is not possible -> error
	if latErr != nil || lngErr != nil {
		// output a not-nil error
		outErr := latErr
		if outErr == nil {
			outErr = lngErr
		}
		w.WriteHeader(http.StatusBadRequest)
		w.Write(common.OutError("Invalid request (lng or lat)", outErr))
		return
	}

	// execute query
	// TODO: "External" Limit?
	rows, queryErr := common.DB().Query(
		`SELECT canteens.id, city_id, 
				canteens.name, cities.name, location,
				(point($1, $2) <@> location)*1.609344 
					as distance
		 FROM canteens
		 INNER JOIN cities ON cities.id=city_id
		 ORDER BY distance
		 LIMIT 5`, v.Get("lat"), v.Get("lng"))

	// check if any error occured while executing the query
	if queryErr != nil {
		w.WriteHeader(http.StatusInternalServerError)
		w.Write(common.OutError("Query failed!", queryErr))
		return
	}
	defer rows.Close()

	// prepare list of canteens
	var canteens []Canteen

	// go through all results and add those to the list
	for rows.Next() {
		var c Canteen
		rowErr := rows.Scan(&c.ID, &c.CityID, &c.Name, &c.CityName,
			&c.GeoLocation, &c.Distance)
		if rowErr != nil {
			w.WriteHeader(http.StatusInternalServerError)
			w.Write(common.OutError("Scanning row failed!", queryErr))
			return
		}
		canteens = append(canteens, c)
	}

	// serialize the user data as json and send out
	w.Write(common.OutResponse(canteens))
}

File: src/api/canteen.rs
use db::DatabaseExt;
use rustless::{Nesting, Namespace};
use std;
// use util::{SmartEndpoint};
use super::util::handle;

#[derive(RustcDecodable, RustcEncodable)]
struct Canteen {
    id: i32,
    name: String,
    city_id: i32,
    city_name: String,
    // GeoLocation common.Location
    distance: f64,
}


pub fn route(ns: &mut Namespace) {

    ns.get("info/:id", |endpoint| {
        handle(endpoint, |client, params| {
            let db = client.app.db();

            let stmt = db.prepare(
                r#"SELECT canteens.id, canteens.name, city_id, cities.name, 
                CAST(-1 AS float8) AS "distance"
                FROM canteens
                INNER JOIN cities ON cities.id=city_id
                WHERE canteens.id=$1"#).unwrap();

            let mut vec = std::vec::Vec::new();
            let id : i32 = try!(params.get("id"));
            for row in stmt.query(&[&id]).unwrap() {
                let canteen = Canteen {
                    id: row.get(0),
                    name: row.get(1),
                    city_id: row.get(2),
                    city_name: row.get(3),
                    distance: row.get(4),
                };
                vec.push(canteen);
            }

            Ok(vec)
        })
    });
}